kp/consumer.go

```diff
@@ -0,0 +1,102 @@
+package kp
+
+import (
+	"log"
+
+	"github.com/Shopify/sarama"
+)
+
+type KPConsumer interface {
+	Setup(sarama.ConsumerGroupSession) error
+	Cleanup(sarama.ConsumerGroupSession) error
+	ConsumeClaim(sarama.ConsumerGroupSession, sarama.ConsumerGroupClaim) error
+}
+
+// ConsumerStruct represents a Sarama consumer group consumer
+type ConsumerStruct struct {
+	topic           string
+	deadLetterTopic string
+	ready           chan bool
+	Processor       func(key string, message string, retries int, rawMessage *sarama.ConsumerMessage) error
+	producer        KPProducer
+	retries         int
+}
+
+func NewConsumer(topic string, deadLetterTopic string, retries int, processor func(key string, message string, retries int, rawMessage *sarama.ConsumerMessage) error, producer KPProducer) ConsumerStruct {
+	return ConsumerStruct{
+		ready:           make(chan bool),
+		Processor:       processor,
+		producer:        producer,
+		topic:           topic,
+		deadLetterTopic: deadLetterTopic,
+		retries:         retries,
+	}
+}
+
+func (Consumer *ConsumerStruct) GetReady() chan bool {
+	return Consumer.ready
+}
+
+func (consumer *ConsumerStruct) SetReady(ready chan bool) {
+	consumer.ready = ready
+}
+
+func (consumer *ConsumerStruct) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	close(consumer.ready)
+
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (consumer *ConsumerStruct) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+func (consumer *ConsumerStruct) ProcessMessage(message *sarama.ConsumerMessage) error {
+	unmarshaledMessage, retries, err := UnmarshalStringMessage(string(message.Value))
+	if err != nil {
+		log.Printf("Error unmarshaling message: %v", err)
+
+		return err
+	}
+	if retries >= consumer.retries {
+		log.Printf("Message has exceeded retries, removing from kafka")
+
+		return nil
+	}
+	err = consumer.Processor(string(message.Key), unmarshaledMessage, retries, message)
+	if err != nil {
+		marshaledMessage := MarshalStringMessage(unmarshaledMessage, retries+1)
+		err = consumer.producer.ProduceMessage(consumer.deadLetterTopic, string(message.Key), marshaledMessage)
```

Comment:
I'm trying to understand. If the processor returns an error, it reproduces the message into the dead letter topic with retires?

Reply:
IMO, the flow should be 

                        
<img width="1049" alt="Screen Shot 2565-06-27 at 15 27 41" src="https://user-images.githubusercontent.com/88310863/175894973-9e2135e6-f2c8-47f8-95c6-977d9fa31b5a.png">

Reply:
or

Reply:
<img width="866" alt="Screen Shot 2565-06-27 at 15 31 47" src="https://user-images.githubusercontent.com/88310863/175895799-7fe9aa76-bf56-40ae-8c3f-4d7b1569f2c3.png">

Reply:
its like the second one, but im not listening to deal letter